From 78b09ca272933f606ee449d73a6efba6a5620c7e Mon Sep 17 00:00:00 2001
From: Jarkko Koivikko <jarkko.koivikko@code-q.fi>
Date: Mon, 7 Jun 2021 14:04:27 +0300
Subject: [PATCH 2/3] Fix high CPU utilization caused by key repeat timer

Key repeat timer has been broken since the beginning.

The key repeat timer leaked timer instances while processing the event,
causing the accumulation of unprocessed timer events and high CPU
utilization.

Fix the issue by killing the original timer (long press delay 600 ms)
and starting the key repeat timer (repeat delay 50 ms) once.

[ChangeLog] Fixed high CPU utilization caused by key repeat timer.

Pick-to: 5.12 5.15 6.1 6.2
Fixes: QTBUG-94259
Change-Id: Iff47249db27cda3750f497cb02c1cb642261e032
Reviewed-by: Jarkko Koivikko <jarkko.koivikko@code-q.fi>
(cherry picked from commit 63a944ff12580f2c333a162ecaecd12419a39c10)
---
 src/virtualkeyboard/qvirtualkeyboardinputengine.cpp | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/src/virtualkeyboard/qvirtualkeyboardinputengine.cpp b/src/virtualkeyboard/qvirtualkeyboardinputengine.cpp
index e64ea4eb..5f74c2a7 100644
--- a/src/virtualkeyboard/qvirtualkeyboardinputengine.cpp
+++ b/src/virtualkeyboard/qvirtualkeyboardinputengine.cpp
@@ -716,9 +716,11 @@ void QVirtualKeyboardInputEngine::timerEvent(QTimerEvent *timerEvent)
 {
     Q_D(QVirtualKeyboardInputEngine);
     if (timerEvent->timerId() == d->repeatTimer) {
-        d->repeatTimer = 0;
         d->virtualKeyClick(d->activeKey, d->activeKeyText, d->activeKeyModifiers, true);
-        d->repeatTimer = startTimer(50);
+        if (!d->repeatCount) {
+            killTimer(d->repeatTimer);
+            d->repeatTimer = startTimer(50);
+        }
         d->repeatCount++;
     }
 }
-- 
2.35.1

